src/mscorlib/src/System/String.cs

```diff
@@ -159,27 +159,48 @@ public sealed class String : IComparable, ICloneable, IConvertible, IEnumerable
             Contract.Ensures(Contract.Result<String>() != null);
             Contract.EndContractBlock();
 
-            if (separator == null)
-                separator = String.Empty;
+            IList<String> list = values as IList<String>;
 
-
-            using(IEnumerator<String> en = values.GetEnumerator()) {
-                if (!en.MoveNext())
+            if (list != null) {
+                if (list.Count == 0)
                     return String.Empty;
 
+                if (list.Count == 1)
+                    return list[0] ?? String.Empty;
+
                 StringBuilder result = StringBuilderCache.Acquire();
-                if (en.Current != null) {
-                    result.Append(en.Current);
-                }
+                result.Append(list[0]);
 
-                while (en.MoveNext()) {
+                for (int i = 1; i < list.Count; i++) {
                     result.Append(separator);
-                    if (en.Current != null) {
-                        result.Append(en.Current);
-                    }
-                }            
+                    result.Append(list[i]);
+                }
+
                 return StringBuilderCache.GetStringAndRelease(result);
-            }           
+            }
+            else
+            {
+                using (IEnumerator<String> en = values.GetEnumerator()) {
+                    if (!en.MoveNext())
+                        return String.Empty;
+
+                    String value = en.Current;
+
+                    if (en.MoveNext()) { 
+                        StringBuilder result = StringBuilderCache.Acquire();
+                        result.Append(value);
+
+                        while (en.MoveNext()) {
```

Comment:
Thanks @sharwell , incomplete copy from my test harness. I'll add that later as well as a gist which might make it easier to confirm perf.